interpreter/functions/formatting.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functions
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// FormatCallback is an interface that allows user-defined behavior
+// for formatting clause implementations, as well as argument retrieval.
+type FormatCallback interface {
+	String(ref.Val, string) (string, error)
+	Decimal(ref.Val, string) (string, error)
+	Fixed(*int) func(ref.Val, string) (string, error)
+	Scientific(*int) func(ref.Val, string) (string, error)
+	Binary(ref.Val, string) (string, error)
+	Hex(bool) func(ref.Val, string) (string, error)
+	Octal(ref.Val, string) (string, error)
+	Arg(int64) (ref.Val, error)
+	ArgSize() int64
+}
+
+type clauseImpl func(ref.Val, string) (string, error)
+
+// ParseFormatString takes a format string, FormatCallback implementation, and locale as input,
+// and returns a formatted string if formatting succeeds, or an error otherwise.
```

Comment:
nit: Instead of repeating information already in the function signature (what the args are), focus more on describing what the function does?

Reply:
Right, I've adjusted it more to describe what it does.